🏗 Migrated unsplash setting to up to date convention

naz opened the pull request:
splits https://github.com/TryGhost/Ghost/pull/12615 into "slack" and "unsplash" parts
refs closes TryGhost/Team#331
refs #10318

Note to reviewers: the part that needs thorough review is the migration located in [10-refactor-unsplash-setting.js  file](https://github.com/TryGhost/Ghost/pull/12632/files#diff-523b93c05310cc634755e8fc294838e2c2181112e196931c057ec5eda09aedab). The rest is optional if you feel like having a look :wink: 

TODO:
- [x] handling for v2/v3/v4(canary) APIs
- [x] importer compatibility
- [x] clarify the `Object` return type in v4 api

Change/compatibility table for this PR:

### importer

`unsplash` :heavy_check_mark: import from both boolean AND object

### canary/v4

#### `GET /settings/` (browse)

`unsplash` :heavy_check_mark: present in response **with `boolean` value**

#### `GET /settings/:settingName` (read)

`unsplash` :heavy_check_mark: present in response **with `boolean` value**


#### `PUT /settings/` (edit)

`unsplash` :heavy_check_mark: updates setting, accepts a boolean

### v3  Admin API

#### `GET /settings/` (browse)

`unsplash` :heavy_check_mark: present in response **with `Object` value**

#### `GET /settings/:settingName` (read)

`unsplash` :heavy_check_mark: present in response **with `Object` value**

#### `PUT /settings/` (edit)

`unsplash` :heavy_check_mark: updates setting, accepts either boolean or object

### v2 Admin API

#### `GET /settings/` (browse)

`unsplash` :heavy_check_mark: present in response with object value

#### `GET /settings/:settingName` (read)

`unsplash` :heavy_check_mark: present in response with object value

#### `PUT /settings/` (edit)

`unsplash` :heavy_check_mark: updates setting, accepts object

allouis committed:
🏗 Updated "unsplash" setting handling to match updated format

refs https://github.com/TryGhost/Ghost/issues/10318

- JSON object format used in previous "unsplash" setting was considered an
anti-pattern. Flat structure was extracted out of the "unsplash.isActive" JSON.
- The naming convention uses `amp` as  a precedent (https://github.com/TryGhost/Team/issues/331#issuecomment-658815017)

allouis committed:
Refactored usage of pre v4 "unsplash" setting format

refs https://github.com/TryGhost/Ghost/issues/10318

- API changes introduced:

canary/v4 Admin API
GET /settings/ (browse)

+ "unsplash" present in response as boolean value

GET /settings/:settingName (read)

+ "unsplash" present in response as boolean value

PUT /settings/ (edit)

+ "unsplash" updates setting, accepts ONLY  boolean format

v3  Admin API
GET /settings/ (browse)

+ "unsplash" present in response with object value

GET /settings/:settingName (read)

+ "unsplash" present in response with object value

PUT /settings/ (edit)

+ "unsplash" updates setting, accepts either boolean or object formats

v2 Admin API
GET /settings/ (browse)

+ "unsplash" present in response with object value

GET /settings/:settingName (read)

+ "unsplash" present in response with object value

PUT /settings/ (edit)

+ "unsplash" updates setting, accepts object format

tpatel reviewed (changes requested):
I would either add `irreversible: true` in the config (that would require an update to the helper) or `return Promise.reject();` in `down()` like in https://github.com/TryGhost/Ghost/blob/873d21d9751345fe7c983719a4a52aed1ba11cab/core/server/data/migrations/utils.js#L261

I think it's better than doing nothing so that we can't run this migration twice.

naz commented:
@tpatel good spot! Changed the migration :wink:

tpatel reviewed (approved):
awesome 🙌

allouis commented on core/server/data/schema/default-settings.json:
```diff
@@ -408,8 +408,11 @@
-            "type": "object"
+            "defaultValue": "true",
+            "validations": {
+                "isIn": [["true", "false"]]
```
Do we want to include the `isEmpty` validation too? 
```suggestion
                "isEmpty": false,
                "isIn": [["true", "false"]]
```

naz replied:
Yeah, I think that wouln't hurt and would be more explicit :+1:  Although I'm a bit torn on what difference that would make in most cases where we use `isIn` validation in combination with `isEmpty`. If `isIn` does not specify empty string or `null` as an acceptable value that should mean anything but specified values should fail.

allouis commented on core/server/data/migrations/versions/4.0/10-refactor-unsplash-setting.js:
```diff
@@ -0,0 +1,29 @@
+const {createIrreversibleMigration} = require('../../utils');
+
+module.exports = createIrreversibleMigration(async (knex) => {
+    const unsplashSetting = await knex('settings')
```
Daniel reminded me on another PR, we probs wanna ass a little log message for visibility!

naz replied:
Yeeees, agreed :relaxed:

allouis commented on core/server/data/importer/importers/data/settings.js:
```diff
@@ -90,6 +102,48 @@ class SettingsImporter extends BaseImporter {
+                            : deprecatedSetting.value;
+
+                        this.dataToImport.push({
+                            id: deprecatedSetting.id || ObjectId.generate(),
```
This could cause duplicate `id`'s if the deprecated setting is split into multiple settings

naz replied:
Hm.. :thinking:  Yup, very good point!

allouis commented on core/server/data/importer/importers/data/settings.js:
```diff
@@ -17,6 +18,17 @@ const deprecatedSupportedSettingsMap = {
     ghost_foot: 'codeinjection_foot'
 };
 
+const deprecatedSupportedSettingsOneToManyMap = {
```
I like the generalised approach of this! Though I wonder if it might be a little easier to follow with an object like:
```js
unsplash: [{
  getValue(valueFromDb) {
    return JSON.parse(valueFromDb).isActive;
  },
  key: 'unsplash',
  group: 'unsplash',
  type: 'boolean'
]]
```

naz replied:
I like this notation! It's smaller and agree, it's way more readable :+1:

naz replied:
Tried playing with it a little, but having multiple fields (slack has multiple mappings as well) then there's lots of duplicate code that comes from having to do error handling for `JSON.parse`. Might come up with something nice later on, but for now leaving it as "good enough, does the job" code :sweat_smile:

allouis replied:
Makes perfect sense to me :relaxed:

allouis commented on core/server/data/importer/importers/data/settings.js:
```diff
@@ -90,6 +102,48 @@ class SettingsImporter extends BaseImporter {
+                            value: value,
+                            group: to.group,
+                            type: to.type,
+                            flags: null,
```
probably might as well do `to.flags || null`?

naz replied:
yupp :+1:

allouis reviewed (changes requested):
All looks good, only the comment about duplicate `id`'s is essential

naz committed:
Removed duplicate migration

no issue

- The migration didn't want to go away after rebases combined with renames

naz committed:
Updated unsplash setting migration to be idempotent

refs https://github.com/TryGhost/Ghost/issues/10318

- Previous version of migration was not following the idempotence rule of migrations

naz commented:
@daniellockyer @tpatel @allouis I've updated the migration here to follow idempotence rule. The commit which fixes is is this one - https://github.com/TryGhost/Ghost/pull/12632/commits/0bb7be38623d6d4fcd3d5335ad862dd70f2f9ee5

tpatel reviewed (approved):
The migration looks great 🙌
